lib/client-options.js

```diff
@@ -129,6 +129,61 @@ function extend(baseOptions, userOptions) {
 }
 
 /**
+ * Creates a new instance of query options with the values from the user.
+ * When some values are not defined, it takes the default values from
+ * - {@link ExecutionProfile}.
+ * - {@link QueryOptions} from the default options.
+ * @param {Client} client
+ * @param {QueryOptions|function} userOptions
+ * @param {Function} [rowCallback]
+ * @param {Boolean} [logged]
+ */
+function createQueryOptions(client, userOptions, rowCallback, logged) {
+  var profile =
+    client.profileManager.getProfile(userOptions && userOptions.executionProfile) || utils.emptyObject;
+  // userOptions can be undefined and could be of type function (is an optional parameter)
+  userOptions = (!userOptions || typeof userOptions === 'function') ? utils.emptyObject : userOptions;
+  var defaultQueryOptions = client.options.queryOptions;
+  // Using fixed property names is 2 order of magnitude faster than dynamically shallow clone objects
+  return {
+    autoPage: ifUndefined(userOptions.autoPage, defaultQueryOptions.autoPage),
+    captureStackTrace: ifUndefined(userOptions.captureStackTrace, defaultQueryOptions.captureStackTrace),
+    consistency: ifUndefined3(userOptions.consistency, profile.consistency, defaultQueryOptions.consistency),
+    customPayload: ifUndefined(userOptions.customPayload, defaultQueryOptions.customPayload),
+    executionProfile: profile,
+    fetchSize: ifUndefined(userOptions.fetchSize, defaultQueryOptions.fetchSize),
+    hints: userOptions.hints,
+    logged: ifUndefined(userOptions.logged, logged),
+    pageState: userOptions.pageState,
+    prepare: ifUndefined(userOptions.prepare, defaultQueryOptions.prepare),
+    readTimeout: ifUndefined3(userOptions.readTimeout, profile.readTimeout, defaultQueryOptions.readTimeout),
+    retry: ifUndefined3(userOptions.retry, profile.retry, defaultQueryOptions.retry),
```

Comment:
retry isn't defined in `defaultOptions()` but is rather on `client.options.policy` so if its not defined on the user provided query options or profile it will be undefined.